mqtt/policy/src/core/mod.rs

```diff
@@ -702,25 +709,130 @@ pub(crate) mod tests {
+    fn all_identity_variable_rules_must_be_evaluated_resources_do_not_match() {
+        let json = r###"{
+            "schemaVersion": "2020-10-30",
+            "statements": [
+                {
+                    "effect": "deny",
+                    "identities": [
+                        "{{any}}"
+                    ],
+                    "operations": [
+                        "write"
+                    ],
+                    "resources": [
+                        "hello/b"
+                    ]
+                },
+                {
+                    "effect": "allow",
+                    "identities": [
+                        "{{identity}}"
+                    ],
+                    "operations": [
+                        "write"
+                    ],
+                    "resources": [
+                        "hello/a"
+                    ]
+                }
+            ]
+        }"###;
+
+        let policy = PolicyBuilder::from_json(json)
+            .with_default_decision(Decision::Denied)
+            .with_substituter(TestIdentitySubstituter)
+            .with_matcher(DefaultResourceMatcher)
+            .build()
+            .expect("Unable to build policy from json.");
+
+        let request = Request::new("actor_a", "write", "hello/a").unwrap();
+
+        assert_matches!(policy.evaluate(&request), Ok(Decision::Allowed));
+    }
+
+    /// Scenario:
+    /// - The same as test case above,
+    ///   but statement operations are different.
+    ///
+    /// This case is created as a result of a discovered bug.
+    #[test]
+    fn all_identity_variable_rules_must_be_evaluated_operations_do_not_match() {
+        let json = r###"{
+            "schemaVersion": "2020-10-30",
+            "statements": [
+                {
+                    "effect": "deny",
+                    "identities": [
+                        "{{any}}"
+                    ],
+                    "operations": [
+                        "read"
+                    ],
+                    "resources": [
+                        "hello/b"
+                    ]
+                },
+                {
+                    "effect": "allow",
+                    "identities": [
+                        "{{identity}}"
+                    ],
+                    "operations": [
+                        "write"
+                    ],
+                    "resources": [
+                        "hello/a"
+                    ]
+                }
+            ]
+        }"###;
+
+        let policy = PolicyBuilder::from_json(json)
+            .with_default_decision(Decision::Denied)
+            .with_substituter(TestIdentitySubstituter)
+            .with_matcher(DefaultResourceMatcher)
+            .build()
+            .expect("Unable to build policy from json.");
+
+        let request = Request::new("actor_a", "write", "hello/a").unwrap();
+
+        assert_matches!(policy.evaluate(&request), Ok(Decision::Allowed));
+    }
```

Comment:
nit: we could probably eliminate duplicated code by paramaterizing test.

Reply:
good point, but since json definition is also different, I thought that it wouldn't look that great to have it parametrized via `test_case`. So I just made different tests.